src/finalfusion/io.py

```diff
@@ -0,0 +1,280 @@
+
+    @staticmethod
+    @abstractmethod
+    def chunk_identifier() -> 'ChunkIdentifier':
+        """
+        Get the ChunkIdentifier for this Chunk.
+
+        Returns
+        --------
+        chunk_identifier : ChunkIdentifier
+        """
+    @staticmethod
+    @abstractmethod
+    def read_chunk(file: IO[bytes]) -> 'Chunk':
+        """
+        Read the Chunk and return it.
+
+        The file must be positioned before the contents of the :class:`Chunk`
+        but after its header.
+
+        Parameters
+        -----------
+        file : IO[bytes]
+            a finalfusion file containing the given Chunk
+
+        Returns
+        --------
+        chunk: Chunk
+            The chunk read from the file.
+        """
+    @abstractmethod
+    def write_chunk(self, file: IO[bytes]):
+        """
+        Write the Chunk to a file.
+
+        Parameters
+        ----------
+        file : IO[bytes]
+            Output file for the Chunk
+        """
+    @staticmethod
+    def read_chunk_header(file: IO[bytes]
+                          ) -> Optional[Tuple['ChunkIdentifier', int]]:
+        """
+        Reads the chunk header.
+
+        After successfully reading the header, a tuple containing
+        :class:`.ChunkIdentifier` and and integer specifying the chunk size in
+        bytes are returned.
+
+        Parameters
+        ----------
+        file : IO[bytes]
+            a finalfusion file positioned before a chunk header.
+
+        Returns
+        -------
+        chunk_header : Optional[(ChunkIdentifier, int)]
+            None is returned iff the reader is at EOF.
+
+        Raises
+        ------
+        FinalfusionFormatError
+            If only part of the header could be read.
+        """
+        val = Chunk._read_binary(file, "<IQ")
+        if val is None:
+            return None
+        return ChunkIdentifier(val[0]), val[1]
+
+    @staticmethod
+    def _write_binary(file: IO[bytes], struct_fmt: str, *args):
+        """
+        Helper method to write binary data according to the format string.
+        :param file:
+        :param struct_fmt:
+        :param args:
+        :return:
+        """
+        data = struct.pack(struct_fmt, *args)
+        file.write(data)
+
+    @staticmethod
+    def _read_binary(file: IO[bytes], struct_fmt: str) -> Optional[Tuple[int]]:
+        """
+        Helper method to read binary data from a file according to the format
+        string.
+
+        Parameters
+        ----------
+        file : IO[bytes]
+            Output file
+        struct_fmt : str
+            struct format string
+
+        Returns
+        -------
+        data : Optional[tuple]
+            Returns the unpacked data as a tuple. If **no** data could be read,
+            None is returned
+
+        Raises
+        ------
+        FinalfusionFormatError
+            If data could only be read partially.
+        """
+        size = struct.calcsize(struct_fmt)
+        buf = file.read(size)
```

Comment:
Is there are reason why this is a different case than `len(buf) != size`?

Reply:
When the file is at the end, `file.read `returns an empty buffer, if the file is malformed, a read might only yield part of the expected buffer, then `len(buf)` might be something between 0 and the expected number.  This would result in an exception raised by struct which would complain about being unable to unpack the expected number of bytes from a too small buffer.

E.g. `find_chunk` relies on getting `None` from `read_chunk_header` at EOF.